### DISCUSSION TITLE
XML Render Themes: support scale factors for font size, stroke-width and symbol size

### DISCUSSION THREAD

**Discussion by eddiemuc** [mapsforge/mapsforge#1311] [Q&A] [answered]
In our app [c:geo](https://github.com/cgeo/cgeo) we have user requests to provide user-specific scale factors to be applied on XML Themes to enlarge entities on the map by a user specific value. For example, users would like to use the themes provided by e.g. OpenAndroMaps, Freizeitkarte etc, but they would like to enlarge e.g. displayed symbols by a user-defined percentage. This is mainly due to better readability. Users specifically ask to user-scale these entities:
* Text Sizes
* Stroke Widths (for lines etc)
* Symbol Sizes

Idea is to have one user-provided scaling value for each of the three entities. E.g. one might want to enlarge all Text Sizes by 30% and all Symbol Sizes by 50%

One way to implement this outside of mapsforge would be for us to read the XML Theme Files and rewrite them with enlarged values for e.g. attributes font-size, stroke-width or symbol-height/width. However, we are wondering if there is a better way to make this feature available.

Another way to solve this is if the XML theme providers would already provide and maintain versions with enlarged entities. However, we would need a solution also working for themes where there is no such support provided (our users would be willing to accept certain layout failures resulting from using strange scale factors).

Thus we have several questions:
* Is there a recommendation on how to implement such a feature with mapsforge?
* We found that class `org.mapsforge.map.layer.renderer.TileRendererLayer` already offers setting scale factor for Text Sizes which is doing exactly what we want (method `setTextScale`). One thought would be to extend this with two other methods to set scale factors for stroke-width and symbol-size. What do you think about this? 
  * Would you accept in principal a Pull Request implementing this if I provide it?
* Do you see any other ways to achieve this result?

Thanks a lot and best regards
  EddieMuc

**Comment by devemux86** [comment, marked as answer]
Mapsforge / VTM automatically adapt the tile size based on device dpi.

Additional adjustments are already provided by Mapsforge / VTM.

- Change the text scale: `TileRendererLayer.setTextScale` / `CanvasAdapter.textScale`
- Change the symbol scale: `DisplayModel.symbolScale` / `CanvasAdapter.symbolScale`
- There is no stroke scale, instead and much better:
- Change the tile size: `DisplayModel.setFixedTileSize` / `Tile.SIZE`
- Change the display scale: `DisplayModel.setDefaultUserScaleFactor` / `CanvasAdapter.userScale`

You can test all of these in [Cruiser](https://wiki.openstreetmap.org/wiki/Cruiser).
(Mapsforge / VTM on Android / Desktop)

**Comment by eddiemuc** [comment]
Thanks a lot for the fast answer!
